File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        Node* newHead=NULL,* newTail=NULL;
        Node * curr=head;
        map<Node * ,Node *>mp;
        while(curr)
        {
            Node * newNode = new Node(curr->val);
            if(newHead==NULL)
            newHead=newTail=newNode;
            else 
            newTail=newTail->next=newNode;
            mp[curr] = newNode;
            curr=curr->next;
        }
        curr=head;
        Node * curr2 = newHead;
        while(curr)
        {
            curr2->random = mp[curr->random];
            curr=curr->next;
            curr2=curr2->next;
        }
        return newHead;
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Node curr = head;
        Node prev = null;
        Node res = null;

        Map<Node,Node> mp = new HashMap<>();

        while(curr!=null)
        {
            Node newNode = new Node(curr.val);
            mp.put(curr,newNode);
            if(prev==null) res = newNode;
            else prev.next = newNode;
            prev = newNode;
            curr = curr.next;
        }
        curr = head;
        Node temp = res;

        while(curr!=null)
        {
            temp.random = mp.get(curr.random);
            temp = temp.next;
            curr = curr.next;
        }
        return res;
    }
}